server/src/hardware/protocol/z21/clientkernel.hpp

```diff
@@ -81,13 +84,52 @@ class ClientKernel final : public Kernel
     uint8_t m_firmwareVersionMinor;
     std::function<void(HardwareType, uint8_t, uint8_t)> m_onHardwareInfoChanged;
 
+    /*!
+     * \brief m_trackPowerOn caches command station track power state.
+     *
+     * NOTE: it must be accessed only from event loop thread or from
+     * Z21::ClientKernel::onStart().
+     *
+     * \sa EventLoop
+     */
     TriState m_trackPowerOn;
+
+    /*!
+     * \brief m_emergencyStop caches command station emergency stop state.
+     *
+     * NOTE: it must be accessed only from event loop thread or from
+     * Z21::ClientKernel::onStart().
+     *
+     * \sa EventLoop
+     */
     TriState m_emergencyStop;
     std::function<void(bool)> m_onTrackPowerOnChanged;
     std::function<void()> m_onEmergencyStop;
 
     DecoderController* m_decoderController = nullptr;
 
+    struct LocoCache
+    {
+      enum class Trend : bool
+      {
+          Ascending = 0,
+          Descending
+      };
+
+      uint16_t dccAddress = 0;
+      bool isEStop = false;
+      uint8_t speedStep = 0;
+      uint8_t speedSteps = 0;
+      uint8_t lastReceivedSpeedStep = 0; //Always in 126 steps
+      Trend speedTrend = Trend::Ascending;
+      bool speedTrendExplicitlySet = false;
+      Direction direction = Direction::Unknown;
+      std::chrono::steady_clock::time_point lastSetTime;
+    };
```

Comment:
I don't remember the logic exactly but it's different. Explicitly set is about overwriting existing value or not when receiving new input while Unset would represent a valid value without meaning to calculate trend.
Explicitly set trend disables overwriting for next receive call which disables the flag, allowing future overwriting.
This was needed to avoid interfering with Traintastic throttle acceleration.

Consider this scenario:

1.  Decoder throttle is 50
2. User with external throttle moves it to 100 (Ascending trend). Being it connected directly to Z21, Traintastic has no way to intercept this value before it gets processed.
3. Traintastic gets the feedback from Z21 and updates `decoder->throttle` property
4. If decoder is in a Train, this will trigger Train target speed change and begins acceleration curve
5. Acceleration curve starts at current Train speed so **decoder is put back** at `throttle = 50`
6. This value might seem a descending trend while is just reverting recent throttle update to begin real acceleration
7. Z21 sees new throttle 50, this is unavoidable and might cause problems with sound decoder as they will receive in rapid succession values 50, **100**, 50, 51, 52, ..., 100. There is a spike at the beginning.

It may sound complicated but it's not. If you have doubts, just tell me and I'll explain it better

Reply:
I've made this trend logic to distinguish feedback of our own triggered changes from genuine user throttle changes.
This method is not perfect and will detect false positives leading to user throttle changes being silently ignored.
The explicitly set logic mitigates Train acceleration algorithm from messing all up

Reply:
Ok, we keep it like it is :)